File: resource_diagnostics_updater/resource_diagnostics_updater/resource_diagnostics_updater.py
from diagnostic_msgs.msg import DiagnosticStatus, KeyValue
from rclpy.callback_groups import ReentrantCallbackGroup
from rclpy.qos import qos_profile_sensor_data

from resource_diagnostics_updater.diagnosed_resource import DiagnosedResource
from resource_diagnostics_updater.diagnostics_publisher import DiagnosticsPublisher
from resource_monitoring_interfaces.msg import Field, Resource


class ResourceDiagnosticsUpdater:

    def __init__(
        self,
        diagnostics_publisher: DiagnosticsPublisher,
        diagnosed_resource: DiagnosedResource,
        reentrant_callback_group: ReentrantCallbackGroup,
    ) -> None:

        self.diagnostics_publisher = diagnostics_publisher
        self.diagnosed_resource = diagnosed_resource

        # Create diagnostic status for this resource
        self.diagnostic_status = DiagnosticStatus()
        self.diagnostic_status.name = diagnosed_resource.name
        self.diagnostic_status.message = "unknown"
        self.diagnostic_status.hardware_id = "telegraf_resource_monitor"

        # Register this diagnostic status with the diagnostics publisher so that it gets published
        #  at the publishers timer rate
        self.diagnostics_publisher.add_diagnostic_status(self.diagnostic_status)

        self.diagnostics_publisher.node.create_subscription(
            Resource,
            diagnosed_resource.topic,
            self.resource_callback,
            qos_profile_sensor_data,
            callback_group=reentrant_callback_group,
        )

        self.diagnostics_publisher.node.get_logger().info(
            f"subscribed to resource topic '{diagnosed_resource.topic}' for diagnostics."
        )

    def resource_callback(self, resource_message: Resource) -> None:

        diagnosed_resource_field: Field = next(
            field
            for field in resource_message.fields
            if field.name == self.diagnosed_resource.field
        )

        self.diagnostic_status.values = [
            KeyValue(
                key=diagnosed_resource_field.name,
                value=str(diagnosed_resource_field.value),
            )
        ]

        if diagnosed_resource_field.value >= self.diagnosed_resource.error_threshold:
            self.diagnostic_status.level = DiagnosticStatus.ERROR
            self.diagnostic_status.message = (
                f"error for {self.diagnosed_resource.name} ({self.diagnosed_resource.field}):"
                f" {diagnosed_resource_field.value} over error threshold"
                f" {self.diagnosed_resource.error_threshold}"
            )

        elif diagnosed_resource_field.value >= self.diagnosed_resource.warning_threshold:
            self.diagnostic_status.level = DiagnosticStatus.WARN
            self.diagnostic_status.message = (
                f"warning for {self.diagnosed_resource.name} {self.diagnosed_resource.field}:"
                f" {diagnosed_resource_field.value} over warning threshold"
                f" {self.diagnosed_resource.warning_threshold}"
            )

        else:
            self.diagnostic_status.level = DiagnosticStatus.OK
            self.diagnostic_status.message = f"{self.diagnosed_resource.name} ok"
            return

        # only publish if status is warning or error
        self.publish()

    def publish(self) -> None:
        self.diagnostics_publisher.publish_diagnostics(self.diagnostic_status)

File: resource_diagnostics_updater_cpp/src/resource_diagnostics_updater.cpp

#include <algorithm>
#include <utility>

#include "rclcpp/node.hpp"
#include <diagnostic_msgs/msg/detail/key_value__struct.hpp>

#include <fmt/core.h>
#include <fmt/format.h>
#include <ros2_fmt_logger/logger.hpp>

#include "resource_diagnostics_updater_cpp/diagnosed_resource.hpp"
#include "resource_diagnostics_updater_cpp/diagnostic_publisher.hpp"
#include "resource_diagnostics_updater_cpp/resource_diagnostics_updater.hpp"
#include "resource_monitoring_interfaces/msg/resource.hpp"

using ResourceType = resource_monitoring_interfaces::msg::Resource;
using KeyValueType = diagnostic_msgs::msg::KeyValue;

ResourceDiagnosticsUpdater::ResourceDiagnosticsUpdater(rclcpp::Node& node,
                                                       DiagnosticPublisher& diagnostic_publisher,
                                                       DiagnosedResource&& diagnosed_resource)
  : diagnostic_publisher_{diagnostic_publisher}
  , diagnosed_resource_(std::move(diagnosed_resource))
  , logger_{node.get_logger()}
{
  diagnostic_status_.level = DiagnosticStatusType::STALE;
  diagnostic_status_.name = diagnosed_resource_.name;
  diagnostic_status_.hardware_id = "telegraph_resource_monitor";

  diagnostic_publisher_.add_diagnostic_status(diagnostic_status_);

  subscription_ = node.create_subscription<ResourceType>(
      diagnosed_resource_.topic, 1,
      [this](const ResourceType::ConstSharedPtr& resource) { resource_callback(*resource); });

  logger_.info("resource diagnostic updater initiated for {} in {} on topic {}",
               diagnosed_resource_.name, diagnosed_resource_.field, diagnosed_resource_.topic);
}

void ResourceDiagnosticsUpdater::resource_callback(const ResourceType& resource)
{
  // find returns as an itterator which uses pointer notation
  auto find = std::find_if(
      resource.fields.begin(), resource.fields.end(),
      [this](const auto& field) { return field.name == diagnosed_resource_.field; });

  if (find != resource.fields.end())
  {
    logger_.debug("found target field {} with value {}", find->name, find->value);
  }
  else
  {
    logger_.warn("cannot find {}", diagnosed_resource_.field);
    return;
  }

  KeyValueType key_value{};
  key_value.key = find->name;
  key_value.value = fmt::format("{}", find->value);

  diagnostic_status_.values = {std::move(key_value)};

  if (find->value >= diagnosed_resource_.error_threshold)
  {
    diagnostic_status_.level = DiagnosticStatusType::ERROR;
    diagnostic_status_.message = fmt::format("error for {} ({}): {} over error threshold {}",
                                             diagnosed_resource_.name, diagnosed_resource_.field,
                                             find->value, diagnosed_resource_.error_threshold);
  }
  else if (find->value >= diagnosed_resource_.warning_threshold)
  {
    diagnostic_status_.level = DiagnosticStatusType::WARN;
    diagnostic_status_.message = fmt::format("error for {} ({}): {} over warn threshold {}",
                                             diagnosed_resource_.name, diagnosed_resource_.field,
                                             find->value, diagnosed_resource_.warning_threshold);
  }
  else
  {
    diagnostic_status_.level = DiagnosticStatusType::OK;
    diagnostic_status_.message = fmt::format("{} ok", diagnosed_resource_.name);
    return;
  }
  // immediately publish if at warn or error level
  diagnostic_publisher_.publish_diagnostics({diagnostic_status_});
}